File: src/main/java/com/pubnub/api/models/consumer/access_manager/sum/SpacePermissions.java
package com.pubnub.api.models.consumer.access_manager.sum;

import com.pubnub.api.SpaceId;
import com.pubnub.api.models.consumer.access_manager.v3.PNResource;

public class SpacePermissions extends PNResource<SpacePermissions> {

    private SpacePermissions() {
    }

    public static SpacePermissions id(SpaceId spaceId) {
        SpacePermissions spacePermissions = new SpacePermissions();
        spacePermissions.resourceName = spaceId.getValue();
        return spacePermissions;
    }

    public static SpacePermissions pattern(String spaceIdPattern) {
        SpacePermissions spacePermissions = new SpacePermissions();
        spacePermissions.resourcePattern = spaceIdPattern;
        return spacePermissions;
    }

    @Override
    public SpacePermissions read() {
        return super.read();
    }

    @Override
    public SpacePermissions delete() {
        return super.delete();
    }

    @Override
    public SpacePermissions write() {
        return super.write();
    }

    @Override
    public SpacePermissions get() {
        return super.get();
    }

    @Override
    public SpacePermissions manage() {
        return super.manage();
    }

    @Override
    public SpacePermissions update() {
        return super.update();
    }

    @Override
    public SpacePermissions join() {
        return super.join();
    }

}

File: src/main/kotlin/com/pubnub/api/models/consumer/access_manager/sum/SpacePermissions.kt
package com.pubnub.api.models.consumer.access_manager.sum

import com.pubnub.api.PubNubError
import com.pubnub.api.PubNubException
import com.pubnub.api.SpaceId
import com.pubnub.api.models.consumer.access_manager.v3.ChannelGrant
import com.pubnub.api.models.consumer.access_manager.v3.PNChannelPatternGrant
import com.pubnub.api.models.consumer.access_manager.v3.PNChannelResourceGrant
import com.pubnub.api.models.consumer.access_manager.v3.PNGrant
import com.pubnub.api.models.consumer.access_manager.v3.PNSpacePatternPermissionsGrant
import com.pubnub.api.models.consumer.access_manager.v3.PNSpacePermissionsGrant

interface SpacePermissions : PNGrant {

    companion object {
        fun name(
            spaceId: SpaceId,
            read: Boolean = false,
            write: Boolean = false,
            manage: Boolean = false,
            delete: Boolean = false,
            create: Boolean = false,
            get: Boolean = false,
            join: Boolean = false,
            update: Boolean = false
        ): SpacePermissions = PNSpacePermissionsGrant(
            id = spaceId.value,
            read = read,
            write = write,
            manage = manage,
            delete = delete,
            create = create,
            get = get,
            join = join,
            update = update
        )

        fun pattern(
            pattern: String,
            read: Boolean = false,
            write: Boolean = false,
            manage: Boolean = false,
            delete: Boolean = false,
            create: Boolean = false,
            get: Boolean = false,
            join: Boolean = false,
            update: Boolean = false
        ): SpacePermissions = PNSpacePatternPermissionsGrant(
            id = pattern,
            read = read,
            write = write,
            manage = manage,
            delete = delete,
            create = create,
            get = get,
            join = join,
            update = update
        )
    }
}

fun SpacePermissions.toChannelGrant(): ChannelGrant {
    return when (this) {
        is PNSpacePermissionsGrant -> PNChannelResourceGrant(spacePermissions = this)
        is PNSpacePatternPermissionsGrant -> PNChannelPatternGrant(spacePermissions = this)
        else -> throw PubNubException(pubnubError = PubNubError.INVALID_ARGUMENTS)
    }
}
